From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chad Brubaker <cbrubaker@google.com>
Date: Sun, 20 Dec 2015 20:49:24 -0800
Subject: [PATCH 16/31] DO NOT MERGE Prevent duplicate certificates in
 TrustedCertificateIndex

With the separate caching of intermediate certificates in
TrustManagerImpl a given intermediate may be passed into .index multiple
times. Avoid adding the certificate to the list each time.

(cherry-picked from commit d080e064abba665e2dde2a8c113f837ba4d9a85e)
Bug: 26232830
Change-Id: I6bed2c65d9e42e052b9b1b129200a997e7dca745
(cherry picked from commit 129c0584885effd7f99a22550c0a5f2b789f967e)
(cherry picked from commit 5549787079e6efc433c5020c7ac3800119f117a8)
---
 .../main/java/org/conscrypt/TrustedCertificateIndex.java | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/crypto/src/main/java/org/conscrypt/TrustedCertificateIndex.java b/crypto/src/main/java/org/conscrypt/TrustedCertificateIndex.java
index b322cd191..dbd05068d 100644
--- a/crypto/src/main/java/org/conscrypt/TrustedCertificateIndex.java
+++ b/crypto/src/main/java/org/conscrypt/TrustedCertificateIndex.java
@@ -69,6 +69,15 @@ public final class TrustedCertificateIndex {
             if (anchors == null) {
                 anchors = new ArrayList<TrustAnchor>(1);
                 subjectToTrustAnchors.put(subject, anchors);
+            } else {
+                // Avoid indexing the same certificate multiple times
+                if (cert != null) {
+                    for (TrustAnchor entry : anchors) {
+                        if (cert.equals(entry.getTrustedCert())) {
+                            return;
+                        }
+                    }
+                }
             }
             anchors.add(anchor);
         }
